### PR TITLE
After pivot

### PR DESCRIPTION
**Pull request by tzufgoogle**
An addition of a pivot landmark - 
(1) The pivot is a landmark beyond the goal, on the same path as the goal and not far from it.
(2) The pivot is optional for the sample. That is, not every sample contains a beyond pivot.

-Build, tests and run are working

### CODE REVIEW

File: cabby/geo/map_processing/map_processor.py [hunk 1]
```diff
@@ -40,17 +40,16 @@
...
-    del argv  # Unused.
-    map = map_structure.Map(FLAGS.region, FLAGS.min_s2_level)
+  del argv  # Unused.
+  map = map_structure.Map(FLAGS.region, FLAGS.min_s2_level)
 
-    # Write to disk.
-    if FLAGS.directory is not None:
-      map.write_map(FLAGS.directory)
+  # Write to disk.
+  map.write_map(FLAGS.directory)
```
Comment (jasonbaldridge) [inline cabby/geo/map_processing/map_processor.py:47]: Need to mark 'directory' as required since you no longer check if it is not None?

File: cabby/geo/map_processing/map_structure.py [hunk 2]
```diff
@@ -24,25 +24,22 @@
...
 
 map_logger = logger.create_logger("map.log", 'map')
 
-map_logger = logger.create_logger("map.log", 'map')
-
 
 class Map:
 
     def __init__(self, map_name: Text, level: int, load_directory: Text = None):
         assert map_name == "Manhattan" or map_name == "Pittsburgh" or \
```
Comment (jasonbaldridge) [inline cabby/geo/map_processing/map_structure.py:41]: Use paren instead of '\'

File: cabby/geo/map_processing/map_structure.py [hunk 3]
```diff
@@ -53,29 +50,33 @@ def __init__(self, map_name: Text, level: int, load_directory: Text = None):
...
+                self.polygon_area, network_type='walk')
+            self.nodes, self.edges = ox.graph_to_gdfs(self.nx_graph)
 
+            # Find closest nodes to POI.
+            self.poi['node'] = self.poi['centroid'].apply(lambda x:
+                                                          ox.distance.
+                                                          get_nearest_node(self.
+                                                          nx_graph, util.
+                                                          tuple_from_point(x)))
+
```
Comment (jasonbaldridge) [inline cabby/geo/map_processing/map_structure.py:72]: Rather than breaking this up over multiple lines, it would be better to define a function and apply it here.

File: cabby/geo/map_processing/map_structure.py [hunk 4]
```diff
@@ -53,29 +50,33 @@ def __init__(self, map_name: Text, level: int, load_directory: Text = None):
...
+                                                          get_nearest_node(self.
+                                                          nx_graph, util.
+                                                          tuple_from_point(x)))
+                                                                                
         else:
             self.load_map(load_directory)
         self.create_S2Graph(level)
 
+        self.nodes = self.nodes.set_crs(epsg=4326)
+        self.edges = self.edges.set_crs(epsg=4326)
```
Comment (jasonbaldridge) [inline cabby/geo/map_processing/map_structure.py:78]: Use a global variable that says what 4326 'means' (through its name and associated comment).

File: cabby/geo/map_processing/map_structure.py [hunk 5]
```diff
@@ -98,15 +99,23 @@ def create_S2Graph(self, level: int):
...
         self.poi['cellids'] = self.poi['geometry'].apply(lambda x: util.
-        cellid_from_point(
-            x, level) if isinstance(x, Point) else util.cellid_from_polygon(x, 
-            level))
+                                                         cellid_from_point(
+                                                             x, level) if 
+                                                             isinstance(x, 
+                                                             Point) else util.
+                                                             cellid_from_polygon
+                                                             (x,level))
```
Comment (jasonbaldridge) [inline cabby/geo/map_processing/map_structure.py:107]: As before, write a function -- this is too complex as a lambda expression, which should only be one liners: https://google.github.io/styleguide/pyguide.html#2104-decision

File: cabby/geo/map_processing/map_structure.py [hunk 6]
```diff
@@ -98,15 +99,23 @@ def create_S2Graph(self, level: int):
...
-            x, level) if isinstance(x, Point) else util.cellid_from_polyline(x, 
-            level))
+                                                                 cellid_from_point
+                                                                 (x, level) if 
+                                                                 isinstance(x, 
+                                                                 Point) else 
+                                                                 util.
+                                                                 cellid_from_polyline
+                                                                 (x, level))
+
```
Comment (jasonbaldridge) [inline cabby/geo/map_processing/map_structure.py:118]: Too complex as lambda expression.

File: cabby/geo/map_processing/map_structure.py [hunk 7]
```diff
@@ -134,19 +143,19 @@ def get_valid_path(self, dir_name: Text, name_ending: Text,
...
         base_filename = self.map_name.lower() + name_ending
 
         # Check if directory is valid.
-        assert os.path.exists(dir_name), "Current directory is: {0}. \
-            The directory {1} doesn't exist.".format(os.getcwd(), dir_name)
-        
+        assert os.path.exists(dir_name), "Current directory is: {0}. The \
+        directory {1} doesn't exist.".format(
+            os.getcwd(), dir_name)
```
Comment (jasonbaldridge) [inline cabby/geo/map_processing/map_structure.py:152]: Use parens instead of '\'.

File: cabby/geo/map_processing/map_structure.py [hunk 8]
```diff
@@ -187,6 +196,18 @@ def write_map(self, dir_name: Text):
...
+        path = self.get_valid_path(dir_name, '_edges', '.geojson')
+        if not os.path.exists(path):
+            # Drop columns with list type.
+            self.edges.drop(['highway', 'service', 'oneway', 'lanes', 'bridge',
+                             'access', 'est_width', 'name', 'maxspeed', 'ref'], 
+                             axis=1, inplace=True)
+            self.edges['osmid'] = self.edges['osmid'].apply(lambda x: str(x))
+            self.edges.to_file(path, driver='GeoJSON')
+        else:
+            map_logger.info("path {0} already exist.".format(path))
```
Comment (jasonbaldridge) [inline cabby/geo/map_processing/map_structure.py:209]: path->Path, exist -> exists

Also, preferable to do map_logger.info(f'Path already exists: {path}')

File: cabby/geo/map_processing/map_structure.py [hunk 9]
```diff
@@ -220,3 +241,8 @@ def load_map(self, dir_name: Text):
             path), "path {0} doesn't exist.".format(path)
         self.nodes = gpd.read_file(path, driver='GeoJSON')
 
+        # Load nodes.
+        path = self.get_valid_path(dir_name, '_edges', '.geojson')
+        assert os.path.exists(
+            path), "path {0} doesn't exist.".format(path)
```
Comment (jasonbaldridge) [inline cabby/geo/map_processing/map_structure.py:247]: path -> Path

Preferable: f'Path doesn't exist: {path}'

File: cabby/geo/sample_poi.py [hunk 10]
```diff
@@ -33,13 +33,13 @@
 
 FLAGS = flags.FLAGS
 flags.DEFINE_enum(
-    "region", None, ['Pittsburgh', 'Manhattan'],
-    "Map areas: Manhattan or Pittsburgh.")
+  "region", None, ['Pittsburgh', 'Manhattan'],
+  "Map areas: Manhattan or Pittsburgh.")
```
Comment (jasonbaldridge) [inline cabby/geo/sample_poi.py:37]: Don't need to do anything here, but we clearly have a common set of options involving region and others, so we should probably refactor to pull those common flag definitions rather than redefining them for every application.

File: cabby/geo/walk.py [hunk 11]
```diff
@@ -61,6 +64,17 @@ def compute_route(start_point: Point, end_point: Point, graph: nx.MultiDiGraph,
         print("No route found for the start and end points.")
         return None
     route_nodes = nodes[nodes['osmid'].isin(route)]
+
+    # Create the dictionary that defines the order for sorting according to 
+		#route order.
```
Comment (jasonbaldridge) [inline cabby/geo/walk.py:69]: Fix indentation.

File: cabby/geo/walk.py [hunk 12]
```diff
@@ -107,15 +120,16 @@ def get_start_poi(map: map_structure.Map, end_point: GeoDataFrame) -> \
...
-        map.nx_graph, dest, max_dist=200, weight='length')
+        map.nx_graph, dest, max_dist=400, weight='length')
 
     outer_circle_graph_osmid = list(outer_circle_graph.nodes.keys())
     inner_circle_graph_osmid = list(inner_circle_graph.nodes.keys())
 
     osmid_in_range = [
-        osmid for osmid in outer_circle_graph_osmid if osmid not in inner_circle_graph_osmid]
+        osmid for osmid in outer_circle_graph_osmid if osmid not in 
+				inner_circle_graph_osmid]
```
Comment (jasonbaldridge) [inline cabby/geo/walk.py:132]: Reformat multiline comprehension according to style guide: https://google.github.io/styleguide/pyguide.html#274-decision

(Break at the 'if'.)

File: cabby/geo/walk.py [hunk 13]
```diff
@@ -130,6 +144,7 @@ def get_start_poi(map: map_structure.Map, end_point: GeoDataFrame) -> \
 
     return start_point
 
+
 def get_landmark_if_tag_exists(gdf: GeoDataFrame, tag: Text, main_tag:
                                Text, alt_main_tag: Text) -> GeoDataFrame:
```
Comment (jasonbaldridge) [inline cabby/geo/walk.py:149]: Don't break line at type annotation. (Go ahead and put each arg on its own line if necessary.)

File: cabby/geo/walk.py [hunk 14]
```diff
@@ -164,16 +179,17 @@ def pick_prominent_pivot(df_pivots: GeoDataFrame) -> Optional[GeoDataFrame]:
     '''
 
     tag_pairs = [('wikipedia', 'amenity'), ('wikidata', 'amenity'),
-                 ('brand', 'brand'), ('tourism', 'tourism'), ('tourism', 
-                 'tourism'),('amenity', 'amenity'), ('shop', 'shop')]
+                 ('brand', 'brand'), ('tourism', 'tourism'), 
+								 ('tourism','tourism'), ('amenity', 'amenity'), ('shop', 'shop')
+								 ]
```
Comment (jasonbaldridge) [inline cabby/geo/walk.py:184]: Fix indentation.

File: cabby/geo/walk.py [hunk 15]
```diff
@@ -245,14 +254,114 @@ def get_pivots(route: GeoDataFrame, map: map_structure.Map, end_point:
...
+      route: The route along which a landmark will be chosen.
+    Returns:
+      A single landmark. '''
+
+    if route.shape[0] < 2:
+        # Return Empty.
+        return GeoDataFrame(index=[0], columns=map.nodes.columns).iloc[0]
+
+    last_node_in_route = route.iloc[-1]
+    before_last_node_in_route = route.iloc[-2]
```
Comment (jasonbaldridge) [inline cabby/geo/walk.py:274]: Need to check that route has sufficient length to use these indices?

File: cabby/geo/walk.py [hunk 16]
```diff
@@ -245,14 +254,114 @@ def get_pivots(route: GeoDataFrame, map: map_structure.Map, end_point:
...
+
+    # route_nodes = nodes[nodes['osmid'].isin(route['osmid'].tolist())]
+
+    street_osmid = map.edges[(map.edges['u'] == last_node_in_route['osmid']) & (
+        map.edges['v'] == before_last_node_in_route['osmid'])]['osmid'].iloc[0]
```
Comment (jasonbaldridge) [inline cabby/geo/walk.py:279]: Hard to parse this -- can you break it into multiple lines?

File: cabby/geo/walk.py [hunk 17]
```diff
@@ -245,14 +254,114 @@ def get_pivots(route: GeoDataFrame, map: map_structure.Map, end_point:
...
+
+    try:
+        # Change OSMID to key
+        last_line = map.edges[(map.edges['osmid'] == street_osmid) & (
+            last_node_in_route['osmid'] == map.edges['u']) & 
+						(before_last_node_in_route['osmid'] != map.edges['v'])]
```
Comment (jasonbaldridge) [inline cabby/geo/walk.py:285]: Again, this is hard to parse. Please break into multiple lines if possible.

File: cabby/geo/walk.py [hunk 18]
```diff
@@ -245,14 +254,114 @@ def get_pivots(route: GeoDataFrame, map: map_structure.Map, end_point:
...
+
+        # Remove the route area.
+
+        points_route = route['geometry'].tolist()
+
+        poly_route = Polygon(points_route).buffer(0.0001)
+
+        route_to_endpont = Polygon(
+            [last_node_in_route["geometry"], end_point['centroid'], 
+						last_node_in_route["geometry"]]).buffer(0.0001)
```
Comment (jasonbaldridge) [inline cabby/geo/walk.py:314]: Break into multiple lines.

File: cabby/geo/walk.py [hunk 19]
```diff
@@ -245,14 +254,114 @@ def get_pivots(route: GeoDataFrame, map: map_structure.Map, end_point:
...
+
+    if df_pivots.shape[0] == 0:
+        # Return Empty.
+        return GeoDataFrame(index=[0], columns=route.columns).iloc[0]
+
+    beyond_pivot = pick_prominent_pivot(df_pivots)
+
+    if beyond_pivot is None:
+        # Return Empty.
+        return GeoDataFrame(index=[0], columns=route.columns).iloc[0]
```
Comment (jasonbaldridge) [inline cabby/geo/walk.py:336]: Can/should we just return None in these cases?

File: cabby/geo/walk.py [hunk 20]
```diff
@@ -307,11 +417,11 @@ def get_single_sample(map: map_structure.Map) -> Optional[geo_item.
...
-                                       ['name'], 'main_pivot': rvs_path_entity.
-                                       main_pivot
-                                       ['main_tag'], 'near_pivot': 
-                                       rvs_path_entity.near_pivot['main_tag'], 
-                                       'instruction':
+																			 ['name'], 'main_pivot': rvs_path_entity.
+																			 main_pivot['main_tag'], 'near_pivot': 
+																			 rvs_path_entity.near_pivot['main_tag'],
+                                       'beyond_pivot': rvs_path_entity.
+																			 beyond_pivot['main_tag'],  'instruction':
```
Comment (jasonbaldridge) [inline cabby/geo/walk.py:424]: Fix indentation. I'd do it by doing something like:

```
gdf_tags_start = gpd.GeoDataFrame({
    'end': rvs_path_entity.end_point['name'],
    ...
  })
```

**Review by jasonbaldridge — CHANGES_REQUESTED** [review]
Lots of minor formatting stuff and suggesting renaming some variables. (I just skimmed over the files that looked like they were all just changing 4 space tabs to 2.)

(Go ahead and resolve formatting conversations whenever you fix them, no need for me to verify them!)

File: cabby/data/extract_test.py [hunk 21]
```diff
@@ -20,17 +20,17 @@
...
-        two_pnc_plazza = [x for x in output if x['pageid']==expected][0]
-        self.assertEqual(two_pnc_plazza['pageid'], expected)
-        self.assertIn('Two PNC Plaza', two_pnc_plazza['text'])
-        self.assertEqual(two_pnc_plazza['title'], 'Two PNC Plaza')
-        self.assertEqual(two_pnc_plazza['ref_qid'], 'Q7859146')
-        self.assertEqual(two_pnc_plazza['ref_instance'], 'skyscraper')
+  def testQueryItems(self):
+    # Test plain Wikipedia query for items
+    output = extract.get_data_by_qid('Q7859146')
+    expected = 12717774 #  Two PNC Plaza.
```
[the remark this answers is not in the document]
Comment (tzufgoogle) [reply cabby/data/extract_test.py:26]: changes to expected_pageid

**Review by tzufgoogle — COMMENTED** [review]
Indentation problems - some of them do not apear in VS code.

Build, test and run working

File: cabby/geo/geo_item.py [hunk 22]
```diff
@@ -20,30 +20,35 @@
...
+                 near_pivot, beyond_pivot):
+    """Construct an Entity from the start and end points, route, and pivots.
+    """
+    return GeoEntity(
+      start,
+      end,
+      route,
+      main_pivot,
+      near_pivot,
+			beyond_pivot
```
[the remark this answers is not in the document]
Comment (jasonbaldridge) [reply cabby/geo/geo_item.py]: Yeah, editors can be that way sometimes. It's okay to submit as is after you tried fixing it in editor and we'll chase it down if it still shows up oddly after pushing the code.

File: cabby/geo/walk.py (hunk 19 above)
Comment (jasonbaldridge) [reply cabby/geo/walk.py]: Okay, sounds good.

File: cabby/geo/walk.py (hunk 20 above)
Comment (jasonbaldridge) [reply cabby/geo/walk.py]: Weird! Okay for now.

**Review by jasonbaldridge — APPROVED** [review]
LGTM
